Fix bug in put/get of non-continuous Mat

komakai opened the pull request:
### Pull Request Readiness Checklist

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is a test

Fix for #19502

asmorkalov commented on modules/core/misc/java/test/MatTest.java:
```diff
@@ -455,6 +455,20 @@ public void testGetIntIntShortArray() {
+        matNonContinuous.get(new int[] { 0, 0, 0 }, outNonContinuous);
+        short[] outContinuous = new short[27];
+        matContinuous.get(new int[] { 0, 0, 0 }, outContinuous);
+        assertArrayEquals(outNonContinuous, outContinuous);
```
`assertMatEqual` does not check continuity flag. You can try it instead of custom method.

komakai replied:
@asmorkalov not sure what you mean. The bug this PR is fixing is in the output from the `get` function which is an array. So we need to compare the arrays not the Mats.

asmorkalov replied:
You are right. Sorry, missed `get` call in the code.

alalek commented on modules/java/generator/src/cpp/Mat.cpp:
```diff
@@ -2159,16 +2159,27 @@ template<typename T> static int mat_put(cv::Mat* m, int row, int col, int count,
 
+// returns true if final index was reached
+static bool updateIdx(cv::Mat* mat, std::vector<int>& indices, int inc) {
+    for (int index = 0; index < inc; index++) {
+        if (incIdx(mat, indices)) {
+            return true;
```
Increasing by one multiple times is really slow (for HD images `inc` value is up to 1280).
We need "fast path" to handle "always contiguous" last dim (cols).

```
{
    int cols_dim = mat->dims - 1;
    size_t cols = mat->size[cols_dim];
    CV_CheckLE(cols, (size_t)INT_MAX, "");  // 'int' only, size_t type is not supported for indexes
    size_t new_cols_idx = (size_t)indices[cols_dim] + inc;
    // only "cols" idx is allowed to be increased till the end
    // do not allow to jump forward because we should support non-contiguous layouts too (submats)
    CV_CheckLE(new_cols_idx, cols, "");
    if (new_cols_idx == cols)
    {
        indices[cols_dim] = 0;
        for (int dim = cols_dim - 1; dim >= 0; --dim)
        {
            int new_idx = indices[dim] + 1;
            if (new_idx != mat->size[dim])
            {
                indices[dim] = new_idx;
                return false;
            }
            else
            {
                indices[dim] = 0;
                continue;
            }
        }
        return true;
    }
    else
    {
        indices[cols_dim] = (int)new_cols_idx;
        return false;
    }
}

```

komakai replied:
Actually updating the indices doesn't require any iteration - we can just calculate where the new indices will end up.
But I get your point - we can optimize the copying of data if the trailing dimensions are contiguous. Will push a new optimized implementation.

alalek commented on modules/core/misc/java/test/MatTest.java:
```diff
@@ -455,6 +455,20 @@ public void testGetIntIntShortArray() {
         assertEquals(4, bytesNum);
         assertTrue(Arrays.equals(new short[] {340, 341, 0, 0}, buff11));
+
+        Mat m2 = new Mat(new int[]{ 5, 5, 5 }, CvType.CV_16S);
```
> 5, 5, 5

different numbers are preferable (to ensure that we don't mess with "size[]" array order in implementation):
```
5, 6, 8
```

komakai commented on modules/java/test/pure_test/src/org/opencv/test/OpenCVTestCase.java:
```diff
@@ -314,6 +314,15 @@ public static void assertGE(double v1, double v2) {
     }
 
+    public static void assertArrayEquals(short[] ar1, short[] ar2) {
+        if (ar1.length != ar2.length) {
+            fail("Arrays have different sizes.");
```
@asmorkalov this was a copy/paste from another `assertArrayEquals` function - I'll change all of them to be like this

komakai commented:
>  alalek added the pr: needs rebase label 2 days ago

Rebasing on the 3.4 branch

komakai commented:
I spotted another bug in the calculation of elements remaining in the Mat:
```
    int rest = (int)m->elemSize();
    for (int i = 0; i < m->dims; i++) {
        rest *= (m->size[i] - idx[i]);
    }
```

I will convert to draft while I fix this

komakai commented:
OK - I have reworked this a bit.
Basically the `mat_put_idx` and the `mat_get_idx` implementations were just the same except for the direction in which the data was copied - so I have merged the 2 implementations.
Also the `mat_put(row,col)` and the `mat_get(row,col)` were just special cases of `mat_put_idx` and `mat_get_idx` so I modified them to just call to the generic case. So 4 functions are now essentially condensed down to one.

asmorkalov commented:
@komakai Is it ready for review?

komakai commented:
> @komakai Is it ready for review?

@asmorkalov  Yeah (once the builds pass!)

alalek reviewed (approved):
Thank you :+1:
